src/functions.php

```diff
@@ -1,33 +1,118 @@
+    } else {
+        $videogames = $query->fetchAll();
     }
     return $videogames;
 }
 
+function checkImgExists(array $game): string {
+    if (!$game['image']) {
+        return "";
+    } else {
+        return "<img class='card-img' src='{$game['image']}' alt='Image of {$game['name']} cover'>";
+    }
+}
 function displayAllGames(array $games): string
 {
     $result = "";
     foreach ($games as $game) {
+        $img = checkImgExists($game);
         $result .= "<div class='card'>
-        <img class='card-img' src='{$game['image']}' alt='Image of {$game['name']} cover'>
+        {$img}
         <h2>Game: {$game['name']}</h2>
         <p><strong>Genre:</strong> {$game['genre']}</p>
         <p><strong>Platform:</strong> {$game['platform']}</p>
-        <p><strong>Age Rating:</strong> {$game['age']}</p>
+        <p><strong>Age Rating:</strong> {$game['agerating']}</p>
     </div>";
     }
     return $result;
+}
+
+function getTable(string $tablename, PDO $db): array {
+    $query = $db->prepare("SELECT `id`, `{$tablename}` FROM `{$tablename}`");
+    $result = $query->execute();
+    if (!$result) {
+        throw new ErrorException("Couldn't retrieve information from database");
+    } else {
+        return $query->fetchAll();
+    }
+}
+
+function setDropdownOptions(array $table, $column): string {
+    $options= '';
+    foreach($table as $row) {
+        $options .= "<option value=\"{$row['id']}\">{$row[$column]}</option>";
+    }
+    return $options;
+}
+
+function validateTitle(string $title) {
+    $title = trim($title);
+    if (empty($title) || strlen($title) > 100 || !preg_match("/^[a-zA-Z0-9\s:!]+$/", $title)) {
+        return false;
+    }
+    return htmlspecialchars($title, ENT_QUOTES, 'UTF-8');
+}
+
+function validateGenre(string $genre): bool {
+    $valid_genres = ['1', '2', '3', '4', '5'];
+    return in_array($genre, $valid_genres);
+}
+
+function validatePlatform(string $platform): bool {
+    $valid_platforms = ['1', '2', '3', '4', '5', '6', '7', '8'];
+    return in_array($platform, $valid_platforms);
+}
+
+function validateAgeRating(string $age_rating): bool {
+    $valid_age_ratings = ['1', '2', '3', '4', '5'];
+    return in_array($age_rating, $valid_age_ratings);
+}
+
+function addGameToDB(PDO $db): void {
+    if(isset($_POST['game-title']) && isset($_POST['game-genre']) && isset($_POST['game-platform']) && isset($_POST['game-age'])) {
+        $gametitle = validateTitle($_POST['game-title']);
+        if ($gametitle === false) {
+            die("Invalid game title");
+        }
+
+        $gamegenre = $_POST['game-genre'];
+        if (!validateGenre($gamegenre)) {
+            die("Invalid genre");
+        }
+
+        $gameplatform = $_POST['game-platform'];
+        if (!validatePlatform($gameplatform)) {
+            die("Invalid platform");
+        }
+
+        $gameage = $_POST['game-age'];
+        if (!validateAgeRating($gameage)) {
+            die("Invalid Age Rating");
+        }
```

Comment:
I think validations and santisations should happen before you reach this function - this function only 'accidentally' works because `$_POST` is available everywhere. The function is called `addGameToDB` so it should concern itself with adding rather than sanitising. Pass an array of the cleaned-up data into this function when you are ready to add something, so that we don't need to touch $_POST in here, and the aforementioned Evil Developer On Your Team does not have the excuse to make a mistake.

`die()` exits the PHP program - so if we manage to input some bad data, we don't get back the original web page, we just get back the message text. It would be more helpful to still return the app page, perhaps with some helpful validation text, rather than `die`.

Reply:
Done